nntrainer/opencl/opencl_program.cpp

```diff
@@ -41,6 +43,101 @@ bool Program::BuildProgram(cl_device_id device_id,
     return false;
   }
 
+  // saving kernel binary
+  if (!binaryCreated)
+    return GetProgramInfo(device_id);
+
+  return true;
+}
+
+/**
+ * @brief Utility to get program info and save kernel binaries
+ *
+ * @param device_id OpenCL device id
+ * @return true if successful or false otherwise
+ */
+bool Program::GetProgramInfo(cl_device_id device_id) {
+  // since only one GPU is being used
+  unsigned int num_devices = 1;
+
+  cl_int error_code = CL_SUCCESS;
+
+  // Read the binary size
+  size_t binaries_size[num_devices];
+  error_code =
+    clGetProgramInfo(program_, CL_PROGRAM_BINARY_SIZES,
+                     sizeof(size_t) * num_devices, binaries_size, nullptr);
+
+  if (error_code != CL_SUCCESS) {
+    ml_loge("Failed to get program binary size. OpenCL error code: %d. %s",
+            error_code,
+            (GetProgramBuildInfo(device_id, CL_PROGRAM_BUILD_LOG)).c_str());
+    return false;
+  }
+
+  // Read the kernel name size
+  size_t kernel_names_size;
+  error_code = clGetProgramInfo(program_, CL_PROGRAM_KERNEL_NAMES, 0, nullptr,
+                                &kernel_names_size);
+
+  if (error_code != CL_SUCCESS) {
+    ml_loge("Failed to get program kernel name size. OpenCL error code: %d. %s",
+            error_code,
+            (GetProgramBuildInfo(device_id, CL_PROGRAM_BUILD_LOG)).c_str());
+    return false;
+  }
+
+  // getting the kernel names
+  char kernel_names[kernel_names_size];
+  error_code = clGetProgramInfo(program_, CL_PROGRAM_KERNEL_NAMES,
+                                kernel_names_size, kernel_names, nullptr);
+
+  if (error_code != CL_SUCCESS) {
+    ml_loge("Failed to get program kernel names. OpenCL error code: %d. %s",
+            error_code,
+            (GetProgramBuildInfo(device_id, CL_PROGRAM_BUILD_LOG)).c_str());
+    return false;
+  } else {
+    ml_logi("Saving kernel binary for: %s", std::string(kernel_names).c_str());
+  }
+
+  // Read the binary
+  size_t binaries_ptr_alloc_size = sizeof(unsigned char *) * num_devices;
+  unsigned char *binaries_ptr[num_devices];
+
+  for (unsigned int i = 0; i < num_devices; ++i) {
+    binaries_ptr[i] = new unsigned char[binaries_size[i]];
+  }
+
+  error_code = clGetProgramInfo(program_, CL_PROGRAM_BINARIES,
+                                binaries_ptr_alloc_size, binaries_ptr, nullptr);
+
+  if (error_code != CL_SUCCESS) {
+    ml_loge("Failed to get program binary data. OpenCL error code: %d. %s",
+            error_code,
+            (GetProgramBuildInfo(device_id, CL_PROGRAM_BUILD_LOG)).c_str());
+
+    // cleanup
+    for (unsigned int i = 0; i < num_devices; ++i) {
+      delete[] binaries_ptr[i];
+    }
+    return false;
+  }
+
+  // Write the binary to file
+  // All kernels in the program will be saved in the binary file
+  for (unsigned int i = 0; i < num_devices; ++i) {
+    std::ofstream fs(std::string(kernel_names) + "_kernel.bin",
```

Comment:
It needs to get a parameter save path like `kernel_path + ... ` which can be defined by option.

Reply:
Incorporated opencl-kernel-path pre-processor directive in latest commit